### DISCUSSION TITLE
Installing Espnet for Nvidia Jetson Platform

### DISCUSSION THREAD

**Discussion by shahizat** [espnet/espnet#4570] [Q&A] [answered]
Hi,

Could someone tried to install espnet onto Nvidia Jetson devices, or other arm based architecture devices?

Best regards,
Shakhizat

**Comment by b-flo** [comment, marked as answer]
Hi,

> Could someone tried to install espnet onto Nvidia Jetson devices, or other arm based architecture devices?

It will depend on the task. You only need to modify our pytorch install with a correct wheel/build if there are no dependencies. Otherwise, I can't really comment, we have a lot of tasks.

For pytorch on Nvidia Jetson, you can directly see the [documentation](https://docs.nvidia.com/deeplearning/frameworks/install-pytorch-jetson-platform/index.html#prereqs-install), Section 2.1 has a link for the wheel.
For other ARM targets (v6, v7, and aarch64), there are some repositories available on github.

**Reply by shahizat** [reply]
> Hi,
> 
> > Could someone tried to install espnet onto Nvidia Jetson devices, or other arm based architecture devices?
> 
> It will depend on the task. You only need to modify our pytorch install with a correct wheel/build if there are no dependencies. Otherwise, I can't really comment, we have a lot of tasks.
> 
> For pytorch on Nvidia Jetson, you can directly see the [documentation](https://docs.nvidia.com/deeplearning/frameworks/install-pytorch-jetson-platform/index.html#prereqs-install), Section 2.1 has a link for the wheel. For other ARM targets (v6, v7, and aarch64), there are some repositories available on github.

Hi, 

Thanks for your reply. Please could you please add the support of CUDA 11.4 and pytorch 1.12.0, if it possible.

**Reply by b-flo** [reply]
We'll eventually support new versions but we can't do that without confirming needed changes across tasks.

Note that it does not mean that ESPnet is incompatible with CUDA 11.1+ and torch 1.10+. Depending on the task, you could just manually modify torch and CUDA install and it would work out of the box.

**Reply by shahizat** [reply]
Hi @b-flo , thanks for your reply.

I have torch version of 1.11.0 and CUDA 11.5. Could you please advise how below issue can be solved?

`CUDA_VERSION=11.5
PYTHON=/home/jetson/.virtualenvs/esp-net/bin/python3
PYTHON_VERSION=Python 3.8.10
USE_CONDA=
TH_VERSION=1.11.0
# NOTE(kamo): Temporary fixed numpy version
. ./activate_python.sh && pip install "numpy<=1.21.4"
Requirement already satisfied: numpy<=1.21.4 in /home/jetson/.virtualenvs/esp-net/lib/python3.8/site-packages (1.21.4)
. ./activate_python.sh && ./installers/install_torch.sh "false" "1.11.0" "11.5"
2022-08-17T01:11:49 (install_torch.sh:125:main) [INFO] python_version=3.8.10
2022-08-17T01:11:49 (install_torch.sh:126:main) [INFO] torch_version=1.11.0
2022-08-17T01:11:49 (install_torch.sh:127:main) [INFO] cuda_version=11.5
2022-08-17T01:11:50 (install_torch.sh:86:install_torch) python3 -m pip install torch==1.11.0+cu115 torchaudio==0.11.0+cu115 -f https://download.pytorch.org/whl/torch_stable.html
Looking in links: https://download.pytorch.org/whl/torch_stable.html
ERROR: Could not find a version that satisfies the requirement torch==1.11.0+cu115 (from versions: 1.8.0, 1.8.1, 1.9.0, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1)
ERROR: No matching distribution found for torch==1.11.0+cu115
make: *** [Makefile:95: pytorch.done] Error 1`

**Reply by b-flo** [reply]
You'll need to modify the installation script to use the correct wheel for your JETSON environment.
Our script parse [this list](https://download.pytorch.org/whl/torch_stable.html) to find a wheel containing selected torch version + either selected cuda version or "cpu". However, wheels for JETSON (or ARM) are not formated like that because there is no "cpu" or "cuda" version (e.g.: `torch-1.12.0a0+84d1cb9.nv22.4-cp38-cp38-linux_aarch64.whl`).

**Reply by shahizat** [reply]
> 
Hi, Thanks for your reply, I managed to install espnet using pip3, but I am experiencing another issue with parallel_wavegan package
`Traceback (most recent call last):
  File "synthesize.py", line 47, in <module>
    wav = vocoder.inference(c_mel)
  File "/home/jetson/Projects/ParallelWaveGAN/parallel_wavegan/models/parallel_wavegan.py", line 257, in inference
    c = torch.tensor(c, dtype=torch.float).to(
TypeError: new(): invalid data type 'str'`

**Reply by b-flo** [reply]
Not sure, sorry. @kan-bayashi Could you help on that?

**Reply by kan-bayashi** [reply]
Could you give me the relevant code?

**Reply by shahizat** [reply]
> Could you give me the relevant code?
Hi,  thanks guys for your support. I am using the same code as it was described [here:](https://github.com/IS2AI/Kazakh_TTS/blob/main/tts1/synthesize.py)

Best regard,
Shakhizat

**Reply by shahizat** [reply]
> Could you give me the relevant code?

Without parallel_wavegan and using below code snippet I was able to run it, but voice is not normalized. I can't hear it at all. 

```
tts = Text2Speech.from_pretrained(
    train_config=train_config,
    model_file=model_path, 
    vocoder_file=vocoder_path,
    vocoder_config=vocoder_conf,
    device = "cpu",
    # Only for Tacotron 2 & Transformer
    threshold=0.5,
    # Only for Tacotron 2
    minlenratio=0.0,
    maxlenratio=10.0,
    use_att_constraint=False,
    backward_window=1,
    forward_window=3,
    # Only for FastSpeech & FastSpeech2 & VITS
    speed_control_alpha=1.0,
    # Only for VITS
    noise_scale=0.333,
    noise_scale_dur=0.333,

)
```

**Reply by kan-bayashi** [reply]
Please try `prefer_normalized_feats=True`. 
https://github.com/espnet/espnet/blob/83319022d773c4e6dc075be2b863103d2c48fe97/espnet2/bin/tts_inference.py#L86

**Reply by shahizat** [reply]
Hi @kan-bayashi , thanks for your reply. It did not help. 

I upload my wav file to [here](https://voca.ro/1f6lWCHr3wqN)(https://voca.ro/1f6lWCHr3wqN) . You can check the quality, which is not good.

**Reply by kan-bayashi** [reply]
I tested the following code. It sounds normal (although I could not understand the language).
Please re-check your code.

```bash
# download pretrained model
wget https://issai.nu.edu.kz/wp-content/uploads/2022/03/kaztts_male1_tacotron2_train.loss.ave.zip
wget https://issai.nu.edu.kz/wp-content/uploads/2022/03/parallelwavegan_male1_checkpoint.zip
unzip kaztts_male1_tacotron2_train.loss.ave.zip
unzip parallelwavegan_male1_checkpoint.zip
ipython
```

```python
import soundfile as sf

from espnet2.bin.tts_inference import Text2Speech


model_path = "exp/tts_train_raw_char/train.loss.ave_5best.pth"
vocoder_path = "checkpoint-400000steps.pkl"

tts = Text2Speech.from_pretrained(
    model_file=model_path,
    vocoder_file=vocoder_path,
    device = "cpu",
    threshold=0.5,
    minlenratio=0.0,
    maxlenratio=10.0,
    use_att_constraint=False,
    backward_window=1,
    forward_window=3,
    prefer_normalized_feats=True,
)

wav = tts("Менің атым Қожа болады. Ал сіздің атыңыз қалай?")["wav"]
sf.write("test.wav", wav.numpy(), tts.fs, "PCM_16")
```

[test.wav.zip](https://github.com/espnet/espnet/files/9385866/test.wav.zip)

**Reply by shahizat** [reply]
Hi @kan-bayashi, thanks for your prompt response in this matter. I really appreciate it. i am going to check it today. Quick update, I was able to solve my issue. Correct me if I am wrong below.

If I understood it correctly. There are two main version of espnet, all version above 0.10.0 version use tts = Text2Speech.from_pretrained, but which is below use text2speech = Text2Speech. So how i resolved it, I installed espnet==0.9.6 using pip3 install espnet==0.9.6 with  --no-deps parameter. Then I installed every python dependencies one by one using pip3 except sentencepiece package, which was installed from source for Jetson board. And finally I was able to run it. Could you please clarify, how stats.h5 file is generated, I guess after training using new version?

**Reply by kan-bayashi** [reply]
OK. You used the old version so some options are not available (maybe there is no `prefer_normalized_feats=True` options). `stats.h5` file is always generated when training the neural vocoder but maybe Kazakh_TTS authors did not include it in pretrained model since the text2mel and vocoder used the same statistics for the normalization.

In the current API, there are several patterns:
1. no stats.h5 / prefer_normalized_feats = False
  [text2mel] -> normed feats -> [denormalizer] -> denormed feats -> [vocoder] -> wav
2. no stats.h5 / prefer_normalized_feats = True
  [text2mel] -> normed feats -> [vocoder] -> wav
3. w/ stats.h5 / prefer_normalized_feats = False
    [text2mel] -> normed feats -> [denormalizer] -> denormed feats -> [vocoder normalizer] -> normed feats -> [vocoder] -> wav

Maybe you could not obtain `stats.h5` so you should follow case 2.
